test/topology_experimental_raft/test_topology_recovery_basic.py

```diff
@@ -9,60 +9,103 @@
 import pytest
 import time
 
+from cassandra.policies import WhiteListRoundRobinPolicy
+
 from test.pylib.manager_client import ManagerClient
-from test.pylib.util import wait_for_cql_and_get_hosts
-from test.topology.util import reconnect_driver, enter_recovery_state, \
+from test.pylib.util import wait_for_cql
+from test.topology.util import enter_recovery_state, \
         delete_raft_data_and_upgrade_state, log_run_time, wait_until_upgrade_finishes as wait_until_schema_upgrade_finishes, \
         wait_until_topology_upgrade_finishes, delete_raft_topology_state, wait_for_cdc_generations_publishing, \
         check_system_topology_and_cdc_generations_v3_consistency, start_writes_to_cdc_table, wait_until_last_generation_is_in_use
+from test.topology.conftest import cluster_con
 
 
 @pytest.mark.asyncio
 @log_run_time
 async def test_topology_recovery_basic(request, mode: str, manager: ManagerClient):
     # Increase ring delay to ensure nodes learn about CDC generations before they start operating.
-    cfg = {'ring_delay_ms': 15000 if mode == 'debug' else 5000}
-
-    servers = await manager.servers_add(3, config=cfg)
-    cql = manager.cql
-    assert(cql)
+    ring_delay = 15000 if mode == 'debug' else 5000
+    normal_cfg = {'ring_delay_ms': ring_delay}
+    zero_token_cfg = {'ring_delay_ms': ring_delay, 'join_ring': False}
+
+    servers = [await manager.server_add(config=normal_cfg),
+               await manager.server_add(config=zero_token_cfg),
+               await manager.server_add(config=normal_cfg)]
+
+    # The zero-token node requires a different cql session not to be ignored by the driver because of empty tokens in
+    # the system.peers table.
+    # We need one cql session for both token-owning nodes to continue the write workload during the rolling restart.
+    cql_normal = cluster_con([servers[0].ip_addr, servers[2].ip_addr], 9042, False, load_balancing_policy=
+                             WhiteListRoundRobinPolicy([servers[0].ip_addr, servers[2].ip_addr])).connect()
+    cql_zero_token = cluster_con([servers[1].ip_addr], 9042, False, load_balancing_policy=
+                                 WhiteListRoundRobinPolicy([servers[1].ip_addr])).connect()
+    # In the whole test, cqls[i] and hosts[i] correspond to servers[i].
+    cqls = [cql_normal, cql_zero_token, cql_normal]
+    hosts = [cql_normal.hosts[0], cql_zero_token.hosts[0], cql_normal.hosts[1]]
```

Comment:
Not sure if it's guaranteed that `cql_normal.hosts[i]` will correspond to `servers[i]` for `i in {0, 2}`

Reply:
I did an experiment and it looks like it's guaranteed. I don't know why, and I'd rather not dig into the Python driver internals to understand it. I think I'll just add a preventive swap for the case when `cql_normal.hosts[0]` corresponds to `servers[2]`. It's simple, `cql_normal.hosts[i].address` can be compared with `servers[i].ip_addr`.